runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   }
+
 ///////////////////////////////////////////////////////////////////////////////////////
 // multianewArrayEvaluator:  multi-dimensional new array of objects
 ///////////////////////////////////////////////////////////////////////////////////////
 TR::Register *
 J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    {
-   TR::ILOpCodes opCode = node->getOpCodeValue();
-   TR::Node::recreate(node, TR::acall);
-   TR::Register * targetRegister = directCallEvaluator(node, cg);
-   TR::Node::recreate(node, opCode);
-   return targetRegister;
-   }
+   static char *useDirectHelperCall = feGetEnv("TR_MultiANewArrayEvaluatorUseDirectCall");
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+   TR::Register *targetReg = cg->allocateRegister();
+
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild = node->getThirdChild();
+
+   if (useDirectHelperCall || !secondChild->getOpCode().isLoadConst() || secondChild->getInt()!=2)
+      {
+      traceMsg(cg->comp(), "DCDCDCDC - multianewArrayEvaluator. More than 2 dimensions, so we can't handle this scenario in JITT'ed code\n");
+      return performHelperCall(node, NULL, TR::acall, false, cg);
+      }
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *nonZeroFirstDimLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *oolFailLabel = generateLabelSymbol(cg);
+
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+
+   TR::Register *dimsPtrReg = cg->evaluate(firstChild);
+   TR::Register *dimReg = cg->evaluate(secondChild);
+   TR::Register *classReg = cg->evaluate(thirdChild);
+
+   // First generate the OOL code. This path will be used to handle the scenarios that we don't want to handle in JITT'ed code.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionEnd, cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, oolFailLabel);
+
+   TR::Node *_callNode = createOutlinedCallNode(cg, node, TR::acall);
+   TR::Register *targetReg2 = NULL;
+   targetReg2 = TR::TreeEvaluator::performCall(_callNode, false, cg);
+   cg->decReferenceCount(_callNode);
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, targetReg, targetReg2);
+
+   // After returning from the helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+   // End of OOL
+   outlinedSlowPath->swapInstructionListsWithCompilation();
```

Comment:
We've already evaluated the children in the mainline. Why can't we just do:

```
   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionEnd, cg);
   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
   outlinedSlowPath->swapInstructionListsWithCompilation();

   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, oolFailLabel);

   TR::ILOpCodes opCode = node->getOpCodeValue();
   TR::Node::recreate(node, TR::acall);
   TR::Register* helperCallReturnReg = directCallEvaluator(node, cg);
   TR::Node::recreate(node, opCode);

   generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, targetReg, helperCallReturnReg);
   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
   
   outlinedSlowPath->swapInstructionListsWithCompilation();
```

And remove those two functions created above in this PR.

Reply:
```
   TR::ILOpCodes opCode = node->getOpCodeValue();
   TR::Node::recreate(node, TR::acall);
   TR::Register* helperCallReturnReg = directCallEvaluator(node, cg);
   TR::Node::recreate(node, opCode);
```

This workload is a candidate for refactoring across the two codgens. So that's why I have an extra routine for it. I've made note of this in your comment above about the routine as well. If you don't think we should refactor it then let me know and I'll inline it back in here. Also if we call an evaluator then the reference count of the children will be decremented and we don't want that as we would technically evaluate the same node twice in the evaluator. I can manually increment them before calling the evaluator. Let me know if that is a better solution then creating a new node here.

Reply:
I understand that calling `directCallEvaluator` will decrement the node counts and kill registers potentially (if node reference count reaches 0). We can just move the above snippet of code I pasted to the end of the `multianewArrayEvaluator` and remove the following three lines:

```
   cg->decReferenceCount(node->getFirstChild());
   cg->decReferenceCount(node->getSecondChild());
   cg->decReferenceCount(node->getThirdChild());
```

Instead just put a comment down noting that the call to `directCallEvaluator` in the OOL will have taken care of the reference counts.

Note that this isn't the most ideal situation as ideally we would just call an API which accepts a method symbol and a bunch of argument `TR::Register*` values and generates a call. This is effectively what `directCallEvaluator` does via the linkage classes but in a convoluted way.
 IIRC we don't have a clear API like this today, which is why we recreate the node here to an `acall` and go through the evaluator.

Reply:
The code has been moved to the end of the evaluator, as per your suggestion (see 10d5904).

`performHelperCall` is used to do the bulk of the work, as it's a similar workload.